File: ch36p1079peopleGroupsCoreData/ch36p1079peopleGroupsCoreData/GroupLister.m


#import "GroupLister.h"
#import "PeopleLister.h"
#import "AppDelegate.h"
#import "NSManagedObject+GroupAndPerson.h"

@interface GroupLister () <UIAlertViewDelegate, NSFetchedResultsControllerDelegate>
@property (nonatomic, strong) NSFetchedResultsController* frc;
@end

@implementation GroupLister

- (void) viewDidLoad {
    [super viewDidLoad];
    UIBarButtonItem* b =
    [[UIBarButtonItem alloc]
     initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
     target:self action:@selector(doAdd:)];
    self.navigationItem.rightBarButtonItem = b;
    UIBarButtonItem* b2 =
    [[UIBarButtonItem alloc]
     initWithBarButtonSystemItem:UIBarButtonSystemItemRefresh
     target:self action:@selector(doRefresh:)];
    self.navigationItem.leftBarButtonItem = b;
    self.navigationItem.rightBarButtonItem = b2;
    self.title = @"Groups";
    // no need to register cell; it comes from the storyboard
    // [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"Cell"];
}

// getter for fetched results controller, pretty much boilerplate from template
- (NSFetchedResultsController *)frc {
    if (_frc != nil) {
        return _frc;
    }
    
    NSFetchRequest *req = [[NSFetchRequest alloc] init];
    NSEntityDescription *entity =
    [NSEntityDescription entityForName:@"Group"
                inManagedObjectContext:self.managedObjectContext];
    req.entity = entity;
    req.fetchBatchSize = 20;
    NSSortDescriptor *sortDescriptor =
    [[NSSortDescriptor alloc] initWithKey:@"timestamp" ascending:YES];
    req.sortDescriptors = @[sortDescriptor];
    
    NSFetchedResultsController *afrc =
    [[NSFetchedResultsController alloc]
     initWithFetchRequest:req
     managedObjectContext:self.managedObjectContext
     sectionNameKeyPath:nil
     cacheName:@"Groups"];
    afrc.delegate = self;
    self.frc = afrc;
    
	NSError *error = nil;
	if (![self.frc performFetch:&error]) {
        // Replace this implementation with code to handle the error appropriately.
        // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
	    NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
	    abort();
	}
    
    return _frc;
}


- (void) doRefresh: (id) sender {
    // currently a no-op
}

- (void) doAdd: (id) sender {
    UIAlertView* av = [[UIAlertView alloc] initWithTitle:@"New Group" message:@"Enter name:" delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:@"OK", nil];
    av.alertViewStyle = UIAlertViewStylePlainTextInput;
    [[av textFieldAtIndex:0] setAutocapitalizationType:UITextAutocapitalizationTypeWords];
    [av show];
}
    
-(void)alertView:(UIAlertView *)alertView didDismissWithButtonIndex:(NSInteger)buttonIndex {
    if (buttonIndex == [alertView cancelButtonIndex])
        return;
    NSString* name = [alertView textFieldAtIndex:0].text;
    if (!name)
        return;
    if ([name isEqualToString: @""])
        return;
    NSManagedObjectContext *context = self.frc.managedObjectContext;
    NSEntityDescription *entity = self.frc.fetchRequest.entity;
    NSManagedObject *mo =
    [NSEntityDescription insertNewObjectForEntityForName:entity.name
                                  inManagedObjectContext:context];
    mo.name = name;
    mo.uuid = [[NSUUID UUID] UUIDString];
    mo.timestamp = [NSDate date];
    
    // save context
    NSError *error = nil;
    BOOL ok = [context save:&error];
    if (!ok) {
        NSLog(@"%@", error);
        return;
    }
    PeopleLister* pl = [[PeopleLister alloc] initWithNibName:@"PeopleLister" bundle:nil groupManagedObject:mo];
    [self.navigationController pushViewController:pl animated:YES];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
}


#pragma mark - Table view data source

// all pretty much boilerplate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [self.frc.sections count];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    id<NSFetchedResultsSectionInfo> sectionInfo = self.frc.sections[section];
    return sectionInfo.numberOfObjects;
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    NSManagedObject *object = [self.frc objectAtIndexPath:indexPath];
    cell.textLabel.text = object.name;
    return cell;
}

- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller
{
    [self.tableView reloadData];
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    PeopleLister* pl = [[PeopleLister alloc] initWithNibName:@"PeopleLister" bundle:nil groupManagedObject:[self.frc objectAtIndexPath:indexPath]];
    [self.navigationController pushViewController:pl animated:YES];
}
 

@end

File: iOS10bookexamples/bk2ch23p824peopleGroupsCoreData/ch36p1079peopleGroupsCoreData/GroupLister.swift

import UIKit
import CoreData

class GroupLister: UITableViewController, NSFetchedResultsControllerDelegate {
    
    var managedObjectContext : NSManagedObjectContext!
    
    lazy var frc: NSFetchedResultsController<Group> = {
        let req: NSFetchRequest<Group> = Group.fetchRequest()
        req.fetchBatchSize = 20
        let sortDescriptor = NSSortDescriptor(key:"timestamp", ascending:true)
        req.sortDescriptors = [sortDescriptor]

        let afrc = NSFetchedResultsController(
            fetchRequest:req,
            managedObjectContext:self.managedObjectContext,
            sectionNameKeyPath:nil, cacheName:nil)
        afrc.delegate = self
        do {
            try afrc.performFetch()
        } catch {
            print("Unresolved error \(error)")
            fatalError("Aborting with unresolved error")
        }
        return afrc
    }()


    override func viewDidLoad() {
        super.viewDidLoad()

        let b = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(doAdd))
        self.navigationItem.rightBarButtonItems = [b]
        let b2 = UIBarButtonItem(barButtonSystemItem: .refresh, target: self, action: #selector(doRefresh))
        self.navigationItem.leftBarButtonItems = [b2]
        self.title = "Groups"
        // no need to register cell, comes from storyboard
    }
    
    func doRefresh(_:AnyObject) {
        // currently a no-op
    }
    
    func doAdd(_:AnyObject) {
        let av = UIAlertController(title: "New Group", message: "Enter name:", preferredStyle: .alert)
        av.addTextField {$0.autocapitalizationType = .words}
        av.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        av.addAction(UIAlertAction(title: "OK", style: .default) { _ in
            guard let name = av.textFields![0].text, !name.isEmpty else {return}
            let context = self.frc.managedObjectContext
            let group = Group(context: context)
            group.name = name
            group.uuid = NSUUID().uuidString
            group.timestamp = NSDate()
            
            // save context
            do {
                try context.save()
            } catch {
                print(error)
                return
            }
            let pl = PeopleLister(group: group)
            self.navigationController!.pushViewController(pl, animated: true)
            })
        self.present(av, animated: true)
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return self.frc.sections!.count
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionInfo = self.frc.sections![section]
        return sectionInfo.numberOfObjects
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier:"Cell", for: indexPath)
        cell.accessoryType = .disclosureIndicator
        let group = self.frc.object(at:indexPath)
        cell.textLabel!.text = group.name
        return cell
    }
    
    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        self.tableView.reloadData()
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let pl = PeopleLister(group: self.frc.object(at:indexPath))
        self.navigationController!.pushViewController(pl, animated: true)
    }
 

}
